File: examples/engine/inmemory-go/store.go
// Package main — rat-engine-inmemory-go: a reference `kind: engine` plugin.
//
// Sub-phase-0d reference implementation (ADR-003): its job is to exercise +
// validate the engine/v1 wire contract, not to be a production query engine. It
// is a SELF-CONTAINED in-memory mini-SQL engine — it holds its own tables rather
// than querying a format/storage provider. (The real engine↔format handoff, where
// the engine pulls an Arrow stream from a format plugin via the core gateway, is a
// separate multi-axis integration concern, not a single-axis 0d reference.)
//
// "Bulk" query results that the contract says flow out-of-band as Arrow are, for
// this reference, carried inline through the in-process stream registry
// (stream.go) — the point is to prove the CONTROL-plane contract (Execute / Query
// / Preview), with a stand-in for the data leg, so a second independent impl +
// golden vectors can cross-check behavior.
package main

import "sync"

// row is one record: column -> value. Values are strings (the mini-SQL is
// untyped) to keep the reference trivial and language-neutral.
type row map[string]string

// table is an in-memory relation: ordered columns + insertion-ordered rows.
type table struct {
	cols []string
	rows []row
}

// store is the engine's whole state. Concurrency-safe for the gRPC threadpool. A
// single monotonic snapshot counter is bumped on every mutation and surfaced as
// WriteResult.snapshot_id.
type store struct {
	mu       sync.Mutex
	tables   map[string]*table
	snapshot int64
}

func newStore() *store { return &store{tables: map[string]*table{}} }

// create (re)declares a table with the given columns. CREATE is idempotent-ish:
// re-creating replaces. Returns the new snapshot.
func (s *store) create(name string, cols []string) int64 {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.tables[name] = &table{cols: append([]string(nil), cols...)}
	s.snapshot++
	return s.snapshot
}

// insert binds values positionally to the table's columns and appends a row.
// Returns (rows_affected=1, snapshot) or an error if the table is unknown.
func (s *store) insert(name string, vals []string) (int64, int64, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	t, ok := s.tables[name]
	if !ok {
		return 0, 0, errUnknownTable(name)
	}
	r := row{}
	for i, c := range t.cols {
		if i < len(vals) {
			r[c] = vals[i]
		}
	}
	t.rows = append(t.rows, r)
	s.snapshot++
	return 1, s.snapshot, nil
}

// selectRows applies an optional WHERE equality filter and a projection, returning
// rows in insertion order. Projection ["*"] yields all columns.
func (s *store) selectRows(st stmt) ([]row, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	t, ok := s.tables[st.table]
	if !ok {
		return nil, errUnknownTable(st.table)
	}
	out := make([]row, 0, len(t.rows))
	for _, r := range t.rows {
		if st.hasWhere && r[st.whereCol] != st.whereVal {
			continue
		}
		out = append(out, project(r, st.cols, t.cols))
	}
	return out, nil
}

// project returns a copy of r restricted to the projection columns (or all
// columns when the projection is ["*"]).
func project(r row, proj, allCols []string) row {
	cols := proj
	if len(proj) == 1 && proj[0] == "*" {
		cols = allCols
	}
	out := make(row, len(cols))
	for _, c := range cols {
		if v, ok := r[c]; ok {
			out[c] = v
		}
	}
	return out
}

File: plugins/engine/inmemory-py/store.py
"""In-memory mini-SQL store for rat-engine-inmemory-py — the SECOND independent
`kind: engine` reference (ADR-003).

A from-scratch Python code path (not a port of the Go internals): its job is to
prove the engine/v1 wire contract is implementable by a different implementation in
a different language, passing the SAME golden vectors
(contracts/conformance/engine-v1.json). Self-contained in-memory tables; the real
engine↔format handoff is separate integration work.
"""

import threading
from typing import Dict, List

from sql import SQLError, Stmt

Row = Dict[str, str]


class _Table:
    def __init__(self, cols: List[str]) -> None:
        self.cols = list(cols)
        self.rows: List[Row] = []


class Store:
    """name -> _Table, plus a monotonic snapshot bumped on every mutation."""

    def __init__(self) -> None:
        self._tables: Dict[str, _Table] = {}
        self._snapshot = 0
        self._lock = threading.Lock()

    def create(self, name: str, cols: List[str]) -> int:
        with self._lock:
            self._tables[name] = _Table(cols)
            self._snapshot += 1
            return self._snapshot

    def insert(self, name: str, vals: List[str]) -> tuple[int, int]:
        """Bind values positionally to the table's columns + append. Returns
        (rows_affected=1, snapshot). Raises SQLError on unknown table."""
        with self._lock:
            t = self._tables.get(name)
            if t is None:
                raise SQLError(f"unknown table {name!r}")
            r = {c: vals[i] for i, c in enumerate(t.cols) if i < len(vals)}
            t.rows.append(r)
            self._snapshot += 1
            return 1, self._snapshot

    def select_rows(self, st: Stmt) -> List[Row]:
        """Apply an optional WHERE equality filter + projection, in insertion
        order. Raises SQLError on unknown table."""
        with self._lock:
            t = self._tables.get(st.table)
            if t is None:
                raise SQLError(f"unknown table {st.table!r}")
            out: List[Row] = []
            for r in t.rows:
                if st.has_where and r.get(st.where_col) != st.where_val:
                    continue
                out.append(_project(r, st.cols, t.cols))
            return out


def _project(r: Row, proj: List[str], all_cols: List[str]) -> Row:
    cols = all_cols if proj == ["*"] else proj
    return {c: r[c] for c in cols if c in r}
